Add the first draft of commit module

amaberg opened the pull request:
- Add a first draft of the presentation
- Add some information in the README
- Add a structure for the coming fbf-questions


@dbosk, kan du kika igenom slidesen så att det ser ok ut innan jag spelar in video?

dbosk commented on modules/git/git_commits/slides/contents.tex:
```diff
@@ -0,0 +1,193 @@
+  \begin{itemize}
+      \item Imperative mood, present tense
+      \begin{itemize}
+          \item Should be able to finish the sentence "If applied, this commit will..."
```
Use `\enquote{If applied ...}` instead for quoting.

dbosk commented on modules/git/git_commits/slides/contents.tex:
```diff
@@ -0,0 +1,193 @@
+\end{itemize}
+    
+\end{block}
+    
```
Add an exercise or two that they can do, so that they can practice. For instance, suggest a commit message (subject and body) for a change of some tiny code example. Then do it as a discussion rather than a question with a correct answer.

If you have examples of good and bad commit messages that would also be nice, then they can discuss the good and bad aspects of them. And you can cover them in the video as you would do during a tutorial (övning).

amaberg replied:
I was trying to come up with good examples, but my imagination ran dry 🙈. 

Do you have any example messages that I can use? Especially longer ones were both difficult to find in the repos I have access to and difficult to come up with without actually having done a commit.

dbosk replied:
You can probably find some bad examples here: https://github.com/dbosk/introtools/commits/master :rofl:

dbosk replied:
Some of my former colleagues used to have things like "duh", "meh" as their commit messages.

dbosk reviewed (changes requested):
It's mostly minor changes suggested. The most important "change" (addition) that I'd like to see is adding exercises at the end. Have the exercise on a slide, let the students discuss solutions, then you show your solution when the video continues.

dbosk commented:
Or did you plan to have exercises in another video (with another set of slides)? That would also work.

amaberg commented:
> 

That was the plan! It feels like two distinctly different parts, and there should be less fuss with the filming set-up that way (my father has a big youtube-filming-set-up with green screen, mic, lights and camera that I will be using, but I am not at all familiar with it, so it might take some fiddling)

amaberg commented:
I've now created a first exercise on commit messages (I will add one on commit plans as well), recorded a presentation and a first draft of some fbf-questions. @dbosk , can you please review it thus far?

dbosk commented:
On tis  2 aug 2022 04:43:32, amaberg wrote:
> That was the plan! It feels like two distinctly different parts, and 
> there should be less fuss with the filming set-up that way

One can always cut-and-paste smaller video parts together too ;-)

dbosk commented on modules/git/git_commits/exercise_slides/contents.tex:
```diff
@@ -0,0 +1,120 @@
+\section{Exercise 1 - Commit messages}
+\begin{frame}
+\begin{block}{Sum over list}
+    We wish to add up all non-negative numbers in a list. We have the following piece of code:
```
They haven't learned any programming yet (very first week). But they have covered the terminal, so we can have shell scripts as examples.

dbosk replied:
Hm, they should be able to handle it. You give them everything.

dbosk replied:
Not sure how you want to do with commit plans in the second exercise, if you want to build on this. Then there might be problems.

Counting the number of files in the current working directory and its subdirectories as the example?
```bash
find . -type f | wc -l
```
The bug could be to forget the `-l` option to `wc`. Or miss to add the `-type f` to `find`.

dbosk commented on modules/git/git_commits/exercise_slides/contents.tex:
```diff
@@ -0,0 +1,120 @@
+
+% Make the students write a new (improved) commit message (fbf)
+
+\section{Exercise 2 - Commit plan}
```
What do you want to do for the commit plan?

Give them an assignment (create two or three small shell scripts) and ask them to write a plan for the commits during the development work?

amaberg replied:
Exactly! 

Do you have any ideas for the assignment? I found it difficult to come up with an example that is simple enough for them to understand, but still big enough to warrant multiple commits. Which is why I haven't added that assignment yet

dbosk replied:
I added an assignment now. They should create three scripts. The last of which would require two commits. I also added a question about when it's suitable to branch. (Not obvious to branch for one-commit work, although it is useful too, but at least branch for the last one which requires two commits to be complete.)

Feel free to modify.

amaberg replied:
That looks great! I'll try to record the exercise video today and upload that as well!

dbosk reviewed (approved):
Looks good to me. My biggest "worry" is the for-loop in Java in the exercise.

amaberg commented on modules/git/git_commits/feedbackfruits_questions.md:
```diff
@@ -0,0 +1,11 @@
+|  [1:35](https://youtu.be/YSXv6-aXWkk?t=95)     |  Open-ended question (with answers)   | What are we trying to achieve with our messages?                 |   
+|  [2:45](https://youtu.be/YSXv6-aXWkk?t=165)    |  Mandatory discussion                 | How do we best satisfy these intentions?                         |   
+|  [4:15](https://youtu.be/YSXv6-aXWkk?t=255)    |  Open-ended question (no answers)     | How do we know that the code is functional?                      |   
+|  [7:45](https://youtu.be/YSXv6-aXWkk?t=465)    |  Mandatory discussion                 | Try to come up with an example of a good subject. <br /> After giving your own (unique) example, give feedback on at least 3 other students' examples |
```
I like the idea of committing with an intentionally bad commit message and then having them improve it!

dbosk commented:
Sounds good, great work!

On tor 11 aug 2022 05:05:09, amaberg wrote:
> That looks great! I'll try to record the exercise video today and 
> upload that as well!

amaberg commented:
> 

I didn't get to it yesterday, but it has been recorded and uploaded now

amaberg commented:
> Sounds good, great work!
> […](#)
> On tor 11 aug 2022 05:05:09, amaberg wrote: That looks great! I'll try to record the exercise video today and upload that as well!

I didn't quite get to it yesterday, but it has been recorded and uploaded now

dbosk commented on modules/git/git_commits/README.md:
```diff
@@ -0,0 +1,9 @@
+
+A recording of the presentation can be found [here](https://www.youtube.com/watch?v=YSXv6-aXWkk)
+
+A recording of the exercise-video can be found [here](https://youtu.be/-xt_bEMTssU)
```
I'm still unsure if it will work with the Java for-loop example. Either 
1. explain it more thoroughly and use a pointer to point at and highlight the relevant parts of the code that you're talking about. Or
2. change it to Python which is much closer to human language.

For example:
```python
def sum_positive(num_list):
    sum_list = 0

    for i in range(1, len(num_list)):
        num = num_list[i]
        if num > 0:
            sum_list += num_list[i]

    return sum_list
```

And then run the code, `sum_positive([1, 2, 3])` to show the error. And then run it with the fixed version too, to show the difference the effect the change had.

I think you can just re-record this short part and then edit the pieces together.

dbosk replied:
Or opt for [the terminal based exercise I suggested above](https://github.com/dbosk/introtools/pull/88#discussion_r941211098). Feels like there might be more re-recording though.

dbosk replied:
Maybe that's the best since they should know the terminal by this time. But they might not have had even the first lecture of a programming course.

dbosk replied:
Any updates @amaberg, we're closing in on course start on Monday! :-D

amaberg replied:
I haven't had any time in the last week. I will record the updated version on Saturday. Hopefully that is enough time

dbosk reviewed (changes requested):
You need to either

1. explain the Java-for-loop more (and use a pointer to point at the relevant parts, or at least mention the line numbers).
2. change to Python which is more readable.
3. change to a terminal based example with a bug.

dbosk commented:
Works for me! I've linked the FBF in Canvas, so if you just update the 
FBF with the new video and set the questions it will be fine.

On tor 25 aug 2022 03:08:28, amaberg wrote:
> I haven't had any time in the last week. I will record the updated 
> version on Saturday. Hopefully that is enough time

amaberg commented:
I fixed the exercise slides now and set up a demo-folder on my computer.

I also added all questions on the first presentation to fbf

dbosk commented:
Great, thanks!Sent from a low-usability deviceOn 26 Aug 2022 10:37, amaberg ***@***.***> wrote:
I fixed the exercise slides now and set up a demo-folder on my computer.
I also added all questions on the first presentation to fbf

—Reply to this email directly, view it on GitHub, or unsubscribe.You are receiving this because you were mentioned.Message ID: ***@***.***>

amaberg commented:
Everything should be finished now. I re-recorded exercise 1 (several times, because stuff kept going wrong), edited everything together and uploaded it to youtube (with a new link). 

I also added two more fbf-questions for the exercise-video, added all time stamps in the table. Finally I uploaded the video and all questions to FBF. 

Let me know if there is anything else that I need to fix before monday!

dbosk commented:
I'm very happy with it, this will probably be the best part of the course! :-D

Great work!
